src/frontend/src/lib/components/carousel/Carousel.svelte

```diff
@@ -0,0 +1,235 @@
+<script lang="ts">
+	import { isNullish, nonNullish } from '@dfinity/utils';
+	import { onMount } from 'svelte';
+	import Controls from '$lib/components/carousel/Controls.svelte';
+	import Indicators from '$lib/components/carousel/Indicators.svelte';
+	import {
+		updateSliderFrameVisiblePart,
+		buildCarouselSliderFrame,
+		disableTransition,
+		enableTransition
+	} from '$lib/utils/carousel.utils';
+
+	export let autoplay = 5000;
+	export let duration = 300;
+	export let easing = 'ease-out';
+	export let styleClass: string | undefined = undefined;
+
+	/**
+	 * Carousel container element variables
+	 */
+	let container: HTMLDivElement | undefined;
+	let containerWidth = 0;
+
+	/**
+	 * Computed slider frame element that wraps all slides
+	 */
+	let sliderFrame: HTMLDivElement | undefined;
+
+	/**
+	 * Variables related to the slides
+	 */
+	let slides: Node[];
+	let currentSlide = 0;
+	let totalSlides: number;
+
+	/**
+	 * Autoplay timer
+	 */
+	let autoplayTimer: NodeJS.Timeout | undefined = undefined;
+
+	/**
+	 * Slide transition timer that needed for last-to-first and first-to-last non-animated transform
+	 */
+	let slideTransformTimer: NodeJS.Timeout | undefined = undefined;
+
+	/**
+	 * A const to be used for enabling transition animation
+	 */
+	const animationParams = {
+		easing,
+		duration
+	};
+
+	/**
+	 * Initialise all required data and attach/detach events
+	 */
+	onMount(() => {
+		initializeCarousel();
+		initialiseAutoplayTimer();
+
+		// Resize listener to re-calculate slide frame width
+		window.addEventListener('resize', onResize);
+
+		return () => {
+			clearAutoplayTimer();
+			clearSlideTransformTimer();
+
+			window.removeEventListener('resize', onResize);
+		};
+	});
+
+	/**
+	 * Build slider frame and set required variables
+	 */
+	const initializeCarousel = () => {
+		if (isNullish(container)) {
+			return;
+		}
+
+		// Save this data only on mount
+		if (isNullish(slides)) {
+			slides = [...container.children];
+			totalSlides = slides.length;
+		}
+
+		containerWidth = container.offsetWidth;
```

Comment:
Shouldn't we derive `containerWidth` from the container rather than imperatively setting it's value?

Reply:
Since `container` is initially `undefined`, I added this var to avoid `nonNullish` check in every place where we need this value. Plus, it can only change on mount and on resize (both events will trigger `initializeCarousel`), so the var is always up-to-date. And, finally, it's again a bit shorter than using `container.offsetWidth`.

Reply:
What I mean was:

```
let containerWidth = 0;
$: containerWidth = container?.offsetWidth ?? 0
```

Would a less imperactive approach be interesting here?

Reply:
In this case, I would keep defining containerWidth onMount (right before we generate the frame).

Reply:
oki doki